### PR TITLE
[Feature][starter][spark] Support transform-v2 for spark

### PR DESCRIPTION
**Pull request by FWLamb**
## Purpose of this pull request

refer to https://github.com/apache/incubator-seatunnel/issues/3268.

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

**Comment by FWLamb** [comment]
@hailin0 @john8628 PTAL

**Comment by hailin0** [comment]
great job

**Review by john8628 — DISMISSED** [review]
LGTM

File: seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java [hunk 1]
```diff
@@ -18,44 +18,56 @@
...
-    protected TransformExecuteProcessor(SparkEnvironment sparkEnvironment,
-                                        JobContext jobContext,
-                                        List<? extends Config> pluginConfigs) {
+    protected TransformExecuteProcessor(SparkEnvironment sparkEnvironment, JobContext jobContext, List<? extends Config> pluginConfigs) {
         super(sparkEnvironment, jobContext, pluginConfigs);
     }
 
     @Override
-    protected List<BaseSparkTransform> initializePlugins(List<? extends Config> pluginConfigs) {
-        SeaTunnelSparkTransformPluginDiscovery transformPluginDiscovery = new SeaTunnelSparkTransformPluginDiscovery();
```
Comment (hailin0) [inline seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java:50]: remove this file `org.apache.seatunnel.plugin.discovery.seatunnel.SeaTunnelSparkTransformPluginDiscovery`

File: seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java [hunk 2]
```diff
@@ -70,13 +82,39 @@ public List<Dataset<Row>> execute(List<Dataset<Row>> upstreamDataStreams) throws
...
+            Row row = rowIterator.next();
+            seaTunnelRow = new InternalRowConverter(seaTunnelDataType).reconvert(InternalRow.apply(row.toSeq()));
+            seaTunnelRow = (SeaTunnelRow) transform.map(seaTunnelRow);
+            InternalRow internalRow = new InternalRowConverter(transform.getProducedType()).convert(seaTunnelRow);
+            MutableValue[] mutableValues = ((SpecificInternalRow) internalRow).values();
+            outputRows.add(new GenericRowWithSchema(
+                Arrays.stream(mutableValues).map(MutableValue::boxed).toArray(),
+                structType));
+        }
+        return sparkEnvironment.getSparkSession().createDataFrame(outputRows, structType);
```
Comment (hailin0) [inline seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java]: maybe should use `Dataset<Row> stream = stream.map(...)` or `Dataset<Row> stream = stream.transform(...)`?

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/resources/assertion/fakesource_to_assert.conf [hunk 3]
```diff
@@ -41,8 +41,10 @@ source {
 }
 
 transform {
-  sql {
-    sql = "select name,age from fake"
+  Filter {
+    source_table_name = "fake"
+    result_table_name = "fake1"
+    fields = ["name", "age"]
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/resources/assertion/fakesource_to_assert.conf]: remove

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/resources/assertion/fakesource_to_assert.conf [hunk 4]
```diff
@@ -51,6 +53,7 @@ transform {
 
 sink {
   Assert {
+    source_table_name = "fake1"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/resources/assertion/fakesource_to_assert.conf]: remove

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-datahub-spark-e2e/src/test/resources/datahub/fakesource_to_datahub.conf [hunk 5]
```diff
@@ -42,16 +42,16 @@ source {
...
-      sql = "select name,age from fake"
-    }
-
-  # If you would like to get more information about how to configure seatunnel and see full list of transform plugins,
-  # please go to https://seatunnel.apache.org/docs/transform/sql
+  Filter {
+    source_table_name = "fake"
+    result_table_name = "fake1"
+    fields = ["name", "age"]
+  }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-datahub-spark-e2e/src/test/resources/datahub/fakesource_to_datahub.conf]: remove

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-datahub-spark-e2e/src/test/resources/datahub/fakesource_to_datahub.conf [hunk 6]
```diff
@@ -42,16 +42,16 @@ source {
...
 }
 
 sink {
   DataHub {
+    source_table_name = "fake1"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-datahub-spark-e2e/src/test/resources/datahub/fakesource_to_datahub.conf]: remove

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java [hunk 7]
```diff
@@ -18,14 +18,17 @@
...
 import org.apache.seatunnel.e2e.common.container.TestContainer;
+import org.apache.seatunnel.e2e.common.junit.DisabledOnContainer;
 
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.TestTemplate;
 import org.testcontainers.containers.Container;
 
 import java.io.IOException;
 
+@DisabledOnContainer(value = {}, type = {EngineType.SEATUNNEL, EngineType.FLINK}, disabledReason = "TODO: Transform v2 translation to flink isn't completed")
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java]: revert

**Review by hailin0 — COMMENTED** [review]
waiting for this pr merged https://github.com/apache/incubator-seatunnel/pull/3396

File: seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java (hunk 2 above)
Comment (FWLamb) [reply seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java]: The above method cannot be serialized correctly.
image


Do you think the `Encoders` are right？

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java (hunk 7 above)
Comment (FWLamb) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java]: Do you mean that
@DisabledOnContainer(value = {}, type = {EngineType.SEATUNNEL, EngineType.FLINK, EngineType.SPARK}, disabledReason = "TODO: Transform v2 translation to flink/spark isn't completed")
?

Comment (FWLamb) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java]: I see what you mean, thanks.

File: seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java [hunk 8]
```diff
@@ -70,13 +82,39 @@ public List<Dataset<Row>> execute(List<Dataset<Row>> upstreamDataStreams) throws
...
+        Iterator<Row> rowIterator = stream.toLocalIterator();
+        while (rowIterator.hasNext()) {
+            Row row = rowIterator.next();
+            seaTunnelRow = new InternalRowConverter(seaTunnelDataType).reconvert(InternalRow.apply(row.toSeq()));
+            seaTunnelRow = (SeaTunnelRow) transform.map(seaTunnelRow);
+            InternalRow internalRow = new InternalRowConverter(transform.getProducedType()).convert(seaTunnelRow);
+            MutableValue[] mutableValues = ((SpecificInternalRow) internalRow).values();
+            outputRows.add(new GenericRowWithSchema(
+                Arrays.stream(mutableValues).map(MutableValue::boxed).toArray(),
+                structType));
```
Comment (john8628) [inline seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java:119]: The new transform-v2(Filter Row) should take into account; The row may alert null value;

Comment (john8628) [reply seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java:119]:
```java
private Dataset<Row> sparkTransform2(SeaTunnelTransform transform, Dataset<Row> stream) throws IOException {
    SeaTunnelDataType<?> seaTunnelDataType = TypeConverterUtils.convert(stream.schema());
    transform.setTypeInfo(seaTunnelDataType);
    StructType structType = (StructType) TypeConverterUtils.convert(transform.getProducedType());
    return stream.map((MapFunction<Row, Row>) row -> {
      SeaTunnelRow seaTunnelRow = new InternalRowConverter(seaTunnelDataType).reconvert(InternalRow.apply(row.toSeq()));
      seaTunnelRow = (SeaTunnelRow) transform.map(seaTunnelRow);
      if (seaTunnelRow != null ){
        InternalRow internalRow =  new InternalRowConverter(transform.getProducedType()).convert(seaTunnelRow);
        Row genericRow = new GenericRow(Arrays.stream(((SpecificInternalRow) internalRow).values()).map(MutableValue::boxed).toArray());
        return genericRow;
      }
      return null;
    }, RowEncoder.apply(structType)).filter(Objects::nonNull);
  }
```    


 this code not works well, error msg `org.apache.spark.unsafe.types.UTF8String is not a valid external type for schema of string`

**Comment by hailin0** [comment]
Please rebase from dev branch

File: seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java (hunk 8 above)
Comment (hailin0) [reply seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java:119]: > The new transform-v2(Filter Row) should take into account; The row may alert null value;

@FWLamb

File: seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java [hunk 9]
```diff
@@ -70,13 +82,39 @@ public List<Dataset<Row>> execute(List<Dataset<Row>> upstreamDataStreams) throws
...
+        transform.setTypeInfo(seaTunnelDataType);
+        StructType structType = (StructType) TypeConverterUtils.convert(transform.getProducedType());
+        SeaTunnelRow seaTunnelRow;
+        List<Row> outputRows = new ArrayList<>();
+        Iterator<Row> rowIterator = stream.toLocalIterator();
+        while (rowIterator.hasNext()) {
+            Row row = rowIterator.next();
+            seaTunnelRow = new InternalRowConverter(seaTunnelDataType).reconvert(InternalRow.apply(row.toSeq()));
+            seaTunnelRow = (SeaTunnelRow) transform.map(seaTunnelRow);
+            InternalRow internalRow = new InternalRowConverter(transform.getProducedType()).convert(seaTunnelRow);
```
Comment (hailin0) [inline seatunnel-core/seatunnel-spark-starter/src/main/java/org/apache/seatunnel/core/starter/spark/execution/TransformExecuteProcessor.java]:
```suggestion
        InternalRowConverter inputRowConverter = InternalRowConverter(seaTunnelDataType);
        InternalRowConverter outputRowConverter = InternalRowConverter(transform.getProducedType())
        while (rowIterator.hasNext()) {
            Row row = rowIterator.next();
            seaTunnelRow = inputRowConverter.reconvert(InternalRow.apply(row.toSeq()));
            seaTunnelRow = (SeaTunnelRow) transform.map(seaTunnelRow);
            if (seaTunnelRow == null) {
                continue;
            }
            InternalRow internalRow = outputRowConverter.convert(seaTunnelRow);
```

**Review by hailin0 — DISMISSED** [review]

**Review by hailin0 — APPROVED** [review]

**Review by TyrantLucifer — APPROVED** [review]

**Pull request merged**